File: channel/src/fakeBridge.ts
import { createServer, type IncomingMessage, type ServerResponse } from 'node:http';
import { randomBytes } from 'node:crypto';
import { createInterface } from 'node:readline';
import { pathToFileURL } from 'node:url';
import { TOKEN_HEADER } from './bridge.js';

const MAX_BODY_BYTES = 1_000_000;

export type FakeBridge = {
    readonly url: string;
    readonly token: string;
    readonly resolved: string[][];
    readonly streamTokens: string[];
    readonly resolveTokens: string[];
    readonly push: (batch: unknown) => void;
    readonly pushRaw: (data: string) => void;
    readonly dropStreams: () => void;
    readonly waitForStream: () => Promise<void>;
    readonly close: () => Promise<void>;
};

const readBody = (request: IncomingMessage): Promise<string> =>
    new Promise((resolve, reject) => {
        const parts: Buffer[] = [];
        const state = { size: 0 };
        request.on('data', (chunk: Buffer) => {
            state.size += chunk.length;
            if (state.size > MAX_BODY_BYTES) {
                request.destroy();
                reject(new Error('the body is too large'));
                return;
            }
            parts.push(chunk);
        });
        request.on('end', () => resolve(Buffer.concat(parts).toString('utf8')));
        request.on('error', reject);
    });

const answer = (response: ServerResponse, status: number, body: string): void => {
    response.writeHead(status, { 'content-type': 'text/plain; charset=utf-8' });
    response.end(body);
};

export type FakeBridgeOptions = {
    readonly token?: string;
    readonly onResolve?: (ids: readonly string[]) => void;
};

export const startFakeBridge = async (options: FakeBridgeOptions = {}): Promise<FakeBridge> => {
    const token = options.token ?? randomBytes(24).toString('hex');
    const streams = new Set<ServerResponse>();
    const resolved: string[][] = [];
    const streamTokens: string[] = [];
    const resolveTokens: string[] = [];
    const waiters: (() => void)[] = [];

    const pushRaw = (data: string): void => {
        streams.forEach(stream => stream.write(`data: ${data.replace(/\n/g, '\ndata: ')}\n\n`));
    };

    const server = createServer((request, response) => {
        const given = request.headers[TOKEN_HEADER];
        const path = new URL(request.url ?? '/', 'http://127.0.0.1').pathname;

        if (request.method === 'GET' && path === '/events') {
            streamTokens.push(typeof given === 'string' ? given : '');
            if (given !== token) {
                answer(response, 401, 'the token does not match');
                return;
            }
            response.writeHead(200, {
                'content-type': 'text/event-stream',
                'cache-control': 'no-store',
                connection: 'keep-alive'
            });
            response.write(': open\n\n');
            streams.add(response);
            response.on('close', () => streams.delete(response));
            waiters.splice(0).forEach(waiter => waiter());
            return;
        }

        if (request.method === 'POST' && path === '/resolve') {
            resolveTokens.push(typeof given === 'string' ? given : '');
            if (given !== token) {
                answer(response, 401, 'the token does not match');
                return;
            }
            readBody(request)
                .then(body => {
                    const parsed = JSON.parse(body) as { ids?: unknown };
                    if (!Array.isArray(parsed.ids) || parsed.ids.some(id => typeof id !== 'string')) {
                        answer(response, 400, 'the body must hold an ids array of strings');
                        return;
                    }
                    resolved.push(parsed.ids as string[]);
                    options.onResolve?.(parsed.ids as string[]);
                    answer(response, 200, 'ok');
                })
                .catch(() => answer(response, 400, 'the body did not parse'));
            return;
        }

        answer(response, 404, 'unknown path');
    });

    await new Promise<void>(resolve => server.listen(0, '127.0.0.1', resolve));
    const address = server.address();
    const port = typeof address === 'object' && address !== null ? address.port : 0;

    return {
        url: `http://127.0.0.1:${String(port)}`,
        token,
        resolved,
        streamTokens,
        resolveTokens,
        push: batch => pushRaw(JSON.stringify(batch)),
        pushRaw,
        dropStreams: () => {
            streams.forEach(stream => stream.end());
            streams.clear();
        },
        waitForStream: () =>
            streams.size > 0 ? Promise.resolve() : new Promise<void>(resolve => waiters.push(resolve)),
        close: async () => {
            streams.forEach(stream => stream.destroy());
            streams.clear();
            await new Promise<void>(resolve => server.close(() => resolve()));
        }
    };
};

const runFromStdin = async (): Promise<void> => {
    const fromEnv = process.env['Y_REVIEW_BRIDGE_TOKEN'];
    const bridge = await startFakeBridge({
        ...(fromEnv === undefined ? {} : { token: fromEnv }),
        onResolve: ids => console.error(`the model resolved: ${ids.join(', ')}`)
    });
    console.error('fake bridge running. Put these two values in the MCP configuration file:');
    console.error(`Y_REVIEW_BRIDGE_URL=${bridge.url}`);
    console.error(`Y_REVIEW_BRIDGE_TOKEN=${bridge.token}`);
    console.error('Paste one batch as one line of JSON, then press Enter.');
    createInterface({ input: process.stdin }).on('line', line => {
        if (line.trim().length === 0) {
            return;
        }
        bridge.pushRaw(line.trim());
        console.error('pushed the batch to the channel');
    });
};

if (process.argv[1] !== undefined && pathToFileURL(process.argv[1]).href === import.meta.url) {
    await runFromStdin();
}

File: channel-server/src/test/kotlin/com/yeskiy/yreview/channel/FakeBridge.kt
package com.yeskiy.yreview.channel

import com.sun.net.httpserver.HttpExchange
import com.sun.net.httpserver.HttpServer
import kotlinx.serialization.json.Json
import kotlinx.serialization.json.JsonArray
import kotlinx.serialization.json.JsonPrimitive
import java.io.OutputStream
import java.net.InetAddress
import java.net.InetSocketAddress
import java.nio.charset.StandardCharsets
import java.util.Collections
import java.util.concurrent.Executors

/**
 * The bridge of the IDE, small enough for a test.
 *
 * The server answers the two paths of the contract, keeps every open event stream, and
 * records the token of each request. A test pushes a batch and reads what the channel did.
 */
class FakeBridge(fixedToken: String? = null) {

    val token: String = fixedToken ?: "a-token-of-32-characters-000000000"

    private val streams = Collections.synchronizedList(mutableListOf<HttpExchange>())

    val resolved: MutableList<List<String>> = Collections.synchronizedList(mutableListOf())

    val streamTokens: MutableList<String> = Collections.synchronizedList(mutableListOf())

    val resolveTokens: MutableList<String> = Collections.synchronizedList(mutableListOf())

    val streamKeys: MutableList<String> = Collections.synchronizedList(mutableListOf())

    val resolveKeys: MutableList<String> = Collections.synchronizedList(mutableListOf())

    private val pool = Executors.newCachedThreadPool { runnable ->
        Thread(runnable, "fake-bridge").apply { isDaemon = true }
    }

    private val http: HttpServer =
        HttpServer.create(InetSocketAddress(InetAddress.getByName("127.0.0.1"), 0), 0).apply {
            createContext("/events", ::events)
            createContext("/resolve", ::resolve)
            createContext("/") { exchange -> answer(exchange, 404, "unknown path") }
            executor = pool
            start()
        }

    val url: String = "http://127.0.0.1:${http.address.port}"

    fun streamCount(): Int = streams.size

    fun push(batch: ReviewBatch) = pushRaw(Json.encodeToString(batch))

    /** Writes one Server-Sent Event. A line break inside the value opens a new data line. */
    fun pushRaw(data: String) {
        val frame = "data: ${data.replace("\n", "\ndata: ")}\n\n".toByteArray(StandardCharsets.UTF_8)
        streams.toList().forEach { exchange ->
            runCatching {
                exchange.responseBody.write(frame)
                exchange.responseBody.flush()
            }
        }
    }

    fun dropStreams() {
        streams.toList().forEach { runCatching { it.close() } }
        streams.clear()
    }

    fun waitForStream() = waitFor("the event stream") { streams.isNotEmpty() }

    fun close() {
        dropStreams()
        http.stop(0)
        pool.shutdownNow()
    }

    private fun events(exchange: HttpExchange) {
        streamTokens.add(exchange.requestHeaders.getFirst(BridgeClient.TOKEN_HEADER).orEmpty())
        streamKeys.add(exchange.requestHeaders.getFirst(BridgeClient.SESSION_HEADER).orEmpty())
        if (exchange.requestHeaders.getFirst(BridgeClient.TOKEN_HEADER) != token) {
            return answer(exchange, 401, "the token does not match")
        }
        exchange.responseHeaders.add("content-type", "text/event-stream")
        exchange.sendResponseHeaders(200, 0)
        exchange.responseBody.write(": open\n\n".toByteArray(StandardCharsets.UTF_8))
        exchange.responseBody.flush()
        streams.add(exchange)
    }

    private fun resolve(exchange: HttpExchange) {
        exchange.use {
            resolveTokens.add(exchange.requestHeaders.getFirst(BridgeClient.TOKEN_HEADER).orEmpty())
            resolveKeys.add(exchange.requestHeaders.getFirst(BridgeClient.SESSION_HEADER).orEmpty())
            if (exchange.requestHeaders.getFirst(BridgeClient.TOKEN_HEADER) != token) {
                return answer(exchange, 401, "the token does not match")
            }
            val body = String(exchange.requestBody.readAllBytes(), StandardCharsets.UTF_8)
            val ids = runCatching {
                (Json.parseToJsonElement(body) as kotlinx.serialization.json.JsonObject)["ids"] as JsonArray
            }.getOrNull() ?: return answer(exchange, 400, "the body must hold an ids array of strings")
            resolved.add(ids.map { (it as JsonPrimitive).content })
            answer(exchange, 200, "ok")
        }
    }

    private fun answer(exchange: HttpExchange, status: Int, body: String) {
        val bytes = body.toByteArray(StandardCharsets.UTF_8)
        exchange.responseHeaders.add("content-type", "text/plain; charset=utf-8")
        exchange.sendResponseHeaders(status, bytes.size.toLong())
        exchange.responseBody.use { stream: OutputStream -> stream.write(bytes) }
    }

    companion object {

        private const val DEADLINE_MS = 10_000L

        /** Waits until the check answers true. A test that waits forever helps nobody. */
        fun waitFor(what: String, ready: () -> Boolean) {
            val deadline = System.currentTimeMillis() + DEADLINE_MS
            while (!ready()) {
                check(System.currentTimeMillis() < deadline) { "gave up waiting for $what" }
                Thread.sleep(10)
            }
        }
    }
}
